datawig/imputer.py

```diff
@@ -244,8 +252,169 @@ def fit(self,
+
+        :param label: label value to explain
+        :param k: number of explanations for each input encoder to return
+        :param label_column: name of label column to be explained (optional)
+        """
+
+        if not self.is_explainable:
+            raise ValueError("No explainable data encoders available.")
+
+        # assign label column encoder
+        if label_column is not None:
+            label_encoders = [enc for enc in self.label_encoders if enc.output_column == label_column]
+            if len(label_encoders) == 0:
+                raise ValueError("Could not find label column")
+            else:
+                label_encoder = label_encoders[0]
+        else:
+            label_encoder = self.label_encoders[0]
+
+        # Check whether to-be-explained label value exists.
+        if label not in label_encoder.token_to_idx.keys():
+            raise ValueError("Specified label {} not observed in label encoder".format(label))
+
+        # If k not given return all tokens
+        if k is None:
+            k = int(1e3)
+
+        # assign index of label value (there can be an additional label column for "unobserved" label.
+        label_idx = label_encoder.token_to_idx[label]
+
+        # for each data encoder extract (token_idx, token_idx_correlation_with_label), extract and apply idx2token map.
+        token_weights = {}
+        for encoder, pattern in self.__class_patterns:
+            # extract idx2token mappings
+            if isinstance(encoder, CategoricalEncoder):
+                idx_tuples = zip(pattern[:, label_idx].argsort()[::-1][:k], sorted(pattern[:, label_idx])[::-1][:k])
+                # +1, bc idx_to_token has 0th column for unobserved labels. TODO: what does it do? remove this column!
+                keymap = {i+1: i for i in range(len(encoder.idx_to_token))}
+                idx2token_temp = dict((keymap[key], val) for key, val in encoder.idx_to_token.items())
+            if isinstance(encoder, TfIdfEncoder):
+                idx_tuples = zip(pattern[:, label_idx].argsort()[::-1][:k], sorted(pattern[:, label_idx])[::-1][:k])
+                idx2token_temp = encoder.idx_to_token
+            token_weights[encoder.output_column] = [(idx2token_temp[token], weight) for token, weight in idx_tuples]
+
+        return token_weights
+
+
+    def explain_instance(self, instance: pd.core.series.Series, k: int = 10, label_column: str = None) -> dict:
+        """
+        Return dictionary with entries for each explainable input column of the give instance.
+        Each entry shows the most likely class label for the input and the features with the highest correlation
+        with that class.
+
+        :param instance: row of data frame (or dictionary)
+        :param k: number of explanations (ngrams) for text inputs
+        :param label_column: name of label column to be explained (optional)
+        """
+
+        # assign label column encoder
+        if label_column is not None:
```

Comment:
This looks suspiciously similar to beginning of _explain()_ method.  Could you extract to a method ?

Reply:
good catch.